cpp/src/io/json/json_column.cu

```diff
@@ -0,0 +1,811 @@
+      auto [child_column, names] =
+        json_column_to_cudf_column2(json_col.child_columns.begin()->second,
+                                    d_input,
+                                    options,
+                                    get_child_schema(json_col.child_columns.begin()->first),
+                                    stream,
+                                    mr);
+      column_names.back().children      = names;
+      auto [result_bitmask, null_count] = make_validity(json_col);
+      return {make_lists_column(num_rows,
+                                std::move(offsets_column),
+                                std::move(child_column),
+                                null_count,
+                                std::move(result_bitmask),
+                                stream,
+                                mr),
+              std::move(column_names)};
+      break;
+    }
+    default: CUDF_FAIL("Unsupported column type, yet to be implemented"); break;
+  }
+
+  return {};
+}
+
+table_with_metadata parse_nested_json2(host_span<SymbolT const> input,
+                                       cudf::io::json_reader_options const& options,
+                                       rmm::cuda_stream_view stream,
+                                       rmm::mr::device_memory_resource* mr)
+{
+  CUDF_FUNC_RANGE();
+  auto const new_line_delimited_json = options.is_enabled_lines();
+
+  // Allocate device memory for the JSON input & copy over to device
+  rmm::device_uvector<SymbolT> d_input = cudf::detail::make_device_uvector_async(input, stream);
+
+  // Parse the JSON and get the token stream
+  const auto [tokens_gpu, token_indices_gpu] = get_token_stream(d_input, options, stream);
+  // gpu tree generation
+  auto gpu_tree = get_tree_representation(tokens_gpu, token_indices_gpu, stream);
+  // print_tree(input, gpu_tree, stream);
+
+  auto [gpu_col_id, gpu_row_offsets] = records_orient_tree_traversal(d_input, gpu_tree, stream);
+
+  d_json_column root_column(stream, mr);
+  root_column.type = json_col_t::ListColumn;
+  root_column.child_offsets.resize(2, stream);
+  thrust::fill(rmm::exec_policy(stream),
+               root_column.child_offsets.begin(),
+               root_column.child_offsets.end(),
+               0);
+
+  // Get internal JSON column
+  make_json_column2(d_input, gpu_tree, gpu_col_id, gpu_row_offsets, root_column, stream, mr);
+
+  // data_root refers to the root column of the data represented by the given JSON string
+  auto& data_root =
+    new_line_delimited_json ? root_column : root_column.child_columns.begin()->second;
+
+  // Verify that we were in fact given a list of structs (or in JSON speech: an array of objects)
+  auto constexpr single_child_col_count = 1;
+  CUDF_EXPECTS(data_root.type == json_col_t::ListColumn and
+                 data_root.child_columns.size() == single_child_col_count and
+                 data_root.child_columns.begin()->second.type == json_col_t::StructColumn,
```

Comment:
Does this succeed if you parse an empty array `[]`?

Reply:
It does not. Good catch! updated code with empty columns case.